drivers/src/lms.rs

```diff
@@ -0,0 +1,476 @@
+        n: 24,
+        w: 1,
+        p: 200,
+        ls: 8,
+    },
+    LmotsParameter {
+        algorithm_name: LmotsAlgorithmType::LmotsSha256N24W2,
+        n: 24,
+        w: 2,
+        p: 101,
+        ls: 6,
+    },
+    LmotsParameter {
+        algorithm_name: LmotsAlgorithmType::LmotsSha256N24W4,
+        n: 24,
+        w: 4,
+        p: 51,
+        ls: 4,
+    },
+    LmotsParameter {
+        algorithm_name: LmotsAlgorithmType::LmotsSha256N24W8,
+        n: 24,
+        w: 8,
+        p: 26,
+        ls: 0,
+    },
+];
+
+// maybe this should just return the 5 values and not a struct?
+// similar to how the LMS parameters are returned
+pub fn get_lmots_parameters(
+    algo_type: &LmotsAlgorithmType,
+) -> CaliptraResult<&'static LmotsParameter> {
+    for i in &LMOTS_P {
+        if i.algorithm_name == *algo_type {
+            return Ok(i);
+        }
+    }
+    raise_err!(InvalidLmotsAlgorithmType)
+}
+
+pub fn get_lms_parameters(algo_type: &LmsAlgorithmType) -> CaliptraResult<(u8, u8)> {
+    match algo_type {
+        LmsAlgorithmType::LmsSha256N32H5 => Ok((32, 5)),
+        LmsAlgorithmType::LmsSha256N32H10 => Ok((32, 10)),
+        LmsAlgorithmType::LmsSha256N32H15 => Ok((32, 15)),
+        LmsAlgorithmType::LmsSha256N32H20 => Ok((32, 20)),
+        LmsAlgorithmType::LmsSha256N32H25 => Ok((32, 25)),
+        LmsAlgorithmType::LmsSha256N24H5 => Ok((24, 5)),
+        LmsAlgorithmType::LmsSha256N24H10 => Ok((24, 10)),
+        LmsAlgorithmType::LmsSha256N24H15 => Ok((24, 15)),
+        LmsAlgorithmType::LmsSha256N24H20 => Ok((24, 20)),
+        LmsAlgorithmType::LmsSha256N24H25 => Ok((24, 25)),
+        LmsAlgorithmType::LmsReserved => {
+            raise_err!(InvalidLmsAlgorithmType)
+        }
+    }
+}
+
+// follows pseudo code at https://www.rfc-editor.org/rfc/rfc8554#section-3.1.3
+fn coefficient(s: &[u8], i: usize, w: usize) -> u8 {
+    let blah: u16 = (1 << (w)) - 1;
+    let index = i * w / 8;
+    let b = s[index];
+
+    // extra logic to avoid the divide by 0
+    // which a good compiler would notice only happens when w is 0 and that portion of the
+    // expression could be skipped
+    let mut shift = 8;
+    if w != 0 {
+        shift = 8 - (w * (i % (8 / w)) + w);
+    }
+
+    // Rust errors if we try to shift off all of the bits off from a value
+    // some implementations 0 fill, others do some other filling.
+    // we make this be 0
+    let mut rs = 0;
+    if shift < 8 {
+        rs = b >> shift;
+    }
+    let small_blah = blah as u8;
+    small_blah & rs
+}
+
+fn checksum(algo_type: &LmotsAlgorithmType, input_string: &[u8]) -> CaliptraResult<u16> {
+    let params = get_lmots_parameters(algo_type)?;
+    let mut sum = 0u16;
+    let upper_bound = params.n as u16 * (8 / params.w as u16);
+    for i in 0..upper_bound as usize {
+        sum =
+            sum + ((1 << params.w) - 1) - (coefficient(input_string, i, params.w as usize) as u16);
```

Comment:
nit: this can be a temp variable outside the loop?

Reply:
Done